### PR TITLE
chore(deps): add Python 3.12 support

### PR DESCRIPTION
**Pull request by jeertmans**
## Overview: What does this pull request change?

Currently, Manim can only be installed on Python 3.11 and below. As Python 3.12 just came out, I think it is good time to update a bit Python requirements so that newer Python versions can install Manim.

I updated the dependencies, and test workflows.

## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly

### CODE REVIEW

**Comment by jeertmans** [comment]
Update: some errors seem to be related to NumPy not building https://github.com/python-poetry/poetry/issues/8137.

I tried to enforce `^1.26.0` for NumPy, but it didn't fix :/

See also https://github.com/numpy/numpy/issues/23808.

File: pyproject.toml
Comment (Viicos) [inline pyproject.toml:42]: Is there a reason why these constraints are specified with respect to the Python version?

Comment (jeertmans) [reply pyproject.toml:42]: The may not be the best constraints, but this was needed to make Poetry happy with running `poetry.lock`.

I think this is needed because there is a version of both packages that both support Python 3.8 and 3.12.

Comment (Viicos) [reply pyproject.toml:42]: Hum interesting, I think we might want to have it working without this workaround, as this will affect users installing `manim` as a published package and are not dealing with Poetry at all (i.e. `poetry.lock` is not involved when installing from PyPI)

Comment (jeertmans) [reply pyproject.toml:42]: This should not affect users, since:

- for NumPy, `v1.25` is the first version that does not support Python 3.8, and `v.1.26` is the earliest version supporting Python 3.18;
- for SciPy, the same logic applies.

Comment (jeertmans) [reply pyproject.toml:42]: Made a typo in my previous comment:

> there is **no** version...

Comment (Viicos) [reply pyproject.toml:42]: I'll take a deeper look at this, feels weird having to do this as usually these constraints are solved by the build tool. Anyway, we will soon require `>=1.21` to use `numpy.typing`

Comment (Viicos) [reply pyproject.toml:42]: So I've checked out `main`, deleted the lock file, bumped `python` to `python = ">=3.8,<3.13"` (as you did), and ran `poetry lock` again without issues. Have you done something similar?

Comment (jeertmans) [reply pyproject.toml]: Yes this is what I did (except deleting the lock file), but it didn't work...
Just ran the commands again, and it now works. I have updated the files accordingly.

Comment (Viicos) [reply pyproject.toml]: Nice!

**Comment by behackl** [comment]
I am sort of reluctant to merge this as long as the pipelines fail with the installation of NumPy. 🤔

**Comment by jeertmans** [comment]
@behackl fixed it :-) I just have to update one test that failed, I guess because Python3.12 changed so stuff.

**Comment by jeertmans** [comment]
@behackl this is now fixed.

I had to skip the doctest because https://github.com/python/cpython/pull/101661 changed `OrderedDict`'s `__repr__` (the PR applies to Python 3.12 and Python >= 3.11.4). It should be possible to write doctests conditionally on the Python version, but it seems to be quite cumbersome.

**Comment by behackl** [comment]
@jeertmans thanks for your efforts, this looks good to me!

I'd like to add two more changes to this before we merge it, but no worries; I'll take care of these:
- To avoid straining our API request limit more than strictly required, I'll remove the Python 3.8 CI pipeline. We'll discuss whether we should drop support for Python 3.8 in general for the next release.
- The skipped doctest should be replaced. Actually, I don't see much value in the `get_parameters` function in its current form at all, given that it is just a one-line alias to `inspect.signature`. It should be either removed, or replaced with a function explicitly checking the presence of "dt" as a parameter.

**Comment by jeertmans** [comment]
Thanks @behackl! Feel free to modify the PR, and tell me if I can help!

**Comment by Viicos** [comment]
Regarding the capped version:
- https://iscinumpy.dev/post/bound-version-constraints/#pinning-the-python-version-is-special
- https://iscinumpy.dev/post/poetry-versions/
- https://github.com/hadialqattan/pycln/pull/81#discussion_r733986005

**Comment by jeertmans** [comment]
@Viicos regarding the last link, how would do avoid the error message from Poetry then?

E.g.:

```
➜  manim git:(py3.11-3.12) poetry lock         
Updating dependencies
Resolving dependencies... (10.8s)

The current project's Python requirement (>=3.8,<4) is not compatible with some of the required packages Python requirement:
  - numpy requires Python <3.13,>=3.9, so it will not be satisfied for Python >=3.8,<3.9 || >=3.13,<4
  - numpy requires Python <3.13,>=3.9, so it will not be satisfied for Python >=3.8,<3.9 || >=3.13,<4

Because no versions of numpy match >1.26,<1.26.1 || >1.26.1,<2.0
 and numpy (1.26.0) requires Python <3.13,>=3.9, numpy is forbidden.
So, because numpy (1.26.1) requires Python <3.13,>=3.9
 and manim depends on numpy (^1.26), version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For numpy, a possible solution would be to set the `python` property to ">=3.9,<3.13"
    For numpy, a possible solution would be to set the `python` property to ">=3.9,<3.13"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

**Comment by Viicos** [comment]
> regarding the last link, how would do avoid the error message from Poetry then?

Yes so unfortunately numpy recently pinned their Python version with https://github.com/numpy/numpy/commit/0065919c36b875e9b43d1c9d1ce77a12bbfe825c, without any apparent reason (it is now unpinned on main though).

We are thus blocked by Poetry's behavior now :/

What would be great to have is, as stated in the second article:

> Besides making it doubly important to never cap the Python version, this is due to the clash between specifying lock file dependencies and library dependencies that get propagated to the metadata and therefore PyPI. There might be a great solution to this: Add PEP 621 support, then if both PEP 621’s `project.requires-python` is set and `tool.poetry.dependencies` has a `python` entry, use the former for the project metadata and the latter for the lock file.
> 
> A tool that lets you produce PyPI packages should not force you to set a metadata slot as important as `Requires-Python` based on a lock file you are not even including in the package.

One way to fix this might be to:
- Set the `numpy` requirement to `>=1.24.0` (this is kind of arbitrary)
- Find a way to use `1.24.0` when creating lockfile, so that Poetry doesn't find out about the newly added Python upper limit on numpy's side (I don't know if this is possible).

Anyway, I think it's fine to keep it this way, we will certainly be able to fix this in the future.

**Review by behackl — APPROVED** [review]
I took care of the mentioned changes -- if the pipeline passes, we can merge this. 👍

**Comment by behackl** [comment]
Thanks again, @jeertmans!

**Comment by jeertmans** [comment]
Thanks to both of you @behackl and @Viicos :-)

**Pull request merged**
